File: back/runtime/stackemit/stackemit.go
// Package stackemit is the DP05 emitter: the COMPLETE, deterministic stack
// emission from a content-addressed DAG phase (S23) — the function the
// preview (DP25) and the deploy (DP26) REUSE: they never choose between
// artifacts, they re-emit.
//
//	EmitStack(phase, manifest, ledger, disk) →
//	    {docker-compose.yml, .env.example, start.sh, start_with_rebuild.sh,
//	     traefik.dynamic.yml}
//
// COMPOSITION, NEVER REINVENTION (CLAUDE.md §3, ADR 0007): the compose is
// composeemit.Emit (DP03) verbatim; the env bundle is envemit.Emit (DP04)
// verbatim; the hand-edit gate reuses the S78 regen ledger/disk shapes and
// drift law (records.Hash inequality); the phase verdict reuses phases (S23,
// IsStable computed, never declared); the content addresses reuse records
// (S02, never forked). Only the traefik dynamic config target is NEW here —
// the file-provider projection of the SAME declared topology the DP03 labels
// encode (no new business rule: routers/middleware/service URLs are derived
// from the manifest's declared servers, hosts stay ${VAR} references).
//
// THE CORNERSTONE MIRROR (the whole epic's): EmitStack is a PURE function of
// the phase's content address — same phase → same bytes, N times, on every
// machine (the rapid property pins ∀×2 + Example×100). No clock, no RNG, no
// map-order leak, no I/O.
//
// THE INTERPRETER SIDECAR (ADR 0040 Décision 7, fork E → recommandation A):
// the Go interpreter rides the StackManifest as a role=interpreter service,
// profile core (it always runs), joignable docker_internal — the emission
// carries it into the compose and its port into the .env.example;
// InterpreterSidecar is the pure lookup the Workbench surfaces.
//
// THE WALL (CLAUDE.md §2): re-emission is below-the-line — the phase (source)
// is AUTHORITATIVE, the emitted code regenerable, never the reverse. The
// emitter refuses, fail-closed:
//   - an UNSTABLE phase (PHASE_NOT_STABLE — S23 reused, stability computed);
//   - a manifest NOT PINNED in the phase cut (OUT_OF_SCOPE — EmitStack never
//     emits from an unpinned manifest: the phase decides what is emitted);
//   - a HAND-EDITED gen/ file (EMITTED_FILE_HAND_EDITED — the S78 drift law:
//     the first drift in sorted-path order blocks the WHOLE bundle; no
//     partial re-emission silently overwrites a human edit, §9).
package stackemit

import (
	"sort"
	"strconv"
	"strings"

	"github.com/steph-frtech/aidos/back/archive/phases"
	"github.com/steph-frtech/aidos/back/kernel/records"
	"github.com/steph-frtech/aidos/back/kernel/stackmanifest"
	"github.com/steph-frtech/aidos/back/runtime/blockreason"
	"github.com/steph-frtech/aidos/back/runtime/composeemit"
	"github.com/steph-frtech/aidos/back/runtime/envemit"
	"github.com/steph-frtech/aidos/back/runtime/generators"
	"github.com/steph-frtech/aidos/back/runtime/regen"
)

// TargetTraefikDynamic — the DP05 additive target of the closed kind × target
// matrix: kind stack_manifest × target traefik-dynamic (the file-provider
// projection of the DP03 label topology). Declared, never invented ad hoc.
const TargetTraefikDynamic = "traefik-dynamic"

// KindStackManifest — the source kind this emitter reads (DP02, verbatim).
const KindStackManifest = stackmanifest.KindStackManifest

// targetOrder is the canonical enumeration order of this package's closed
// target set (declared, never derived from map iteration) — DP03/DP04 own
// docker-compose and env-example/start-scripts; stackemit owns only the NEW
// traefik-dynamic target and COMPOSES the rest.
var targetOrder = []string{TargetTraefikDynamic}

// Targets returns every stackemit target in canonical order.
func Targets() []string {
	out := make([]string, len(targetOrder))
	copy(out, targetOrder)
	return out
}

// Bundle is the DP05 emission: the complete deployable stack of ONE phase —
// five byte-stable artifacts (the DP03 compose, the DP04 env bundle, the DP05
// traefik dynamic config), triple-addressed: the phase content address
// (PhaseVersion, S23), the manifest content address (SourceHash, S02) and the
// bundle's own output address (BundleHash — Hash over the canonical artifact
// summary, the «ré-émettre deux fois, hash égaux» proof).
type Bundle struct {
	PhaseVersion     string               `json:"phase_version"`
	SourceHash       string               `json:"source_hash"`
	BundleHash       string               `json:"bundle_hash"`
	Compose          composeemit.Artifact `json:"compose"`
	EnvExample       composeemit.Artifact `json:"env_example"`
	StartSh          composeemit.Artifact `json:"start_sh"`
	StartWithRebuild composeemit.Artifact `json:"start_with_rebuild_sh"`
	TraefikDynamic   composeemit.Artifact `json:"traefik_dynamic"`
}

// Artifacts returns the bundle's five artifacts in canonical order
// (compose, env, start, rebuild, traefik) — the order BundleHash folds over.
func (b Bundle) Artifacts() []composeemit.Artifact {
	return []composeemit.Artifact{
		b.Compose, b.EnvExample, b.StartSh, b.StartWithRebuild, b.TraefikDynamic,
	}
}

// InterpreterSidecar returns the manifest's Go interpreter sidecar service
// (ADR 0040 Décision 7: role=interpreter) and whether one is declared — a
// pure lookup in sorted-name order (deterministic when several are declared).
func InterpreterSidecar(m stackmanifest.StackManifest) (stackmanifest.Service, bool) {
	sorted := make([]stackmanifest.Service, len(m.Services))
	copy(sorted, m.Services)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Name < sorted[j].Name })
	for _, s := range sorted {
		if s.Role == stackmanifest.RoleInterpreter {
			return s, true
		}
	}
	return stackmanifest.Service{}, false
}

// PhaseFor builds the MINIMAL S23 stable phase pinning a manifest: a
// one-constraint cut («stack_manifest/<app>» → the manifest's content
// address), no links, no sensors — vacuously stable (the §43 base case). It
// is the deterministic seed the Workbench and the mirror use; a real build's
// phase comes from the DAG (S23/S96) and carries its full cut — EmitStack
// accepts any stable phase that PINS the manifest.
func PhaseFor(m stackmanifest.StackManifest) (phases.StablePhase, error) {
	h, err := stackmanifest.HashManifest(m)
	if err != nil {
		return phases.StablePhase{}, err
	}
	cut := phases.Cut{KindStackManifest + "/" + m.AppName: h}
	return phases.IsStable(cut, nil, nil, nil), nil
}

// manifestPinned reports whether the manifest's content address appears as a
// pinned version of the phase cut — the phase is the AUTHORITATIVE selection:
// EmitStack never emits from a manifest the phase did not pin.
func manifestPinned(phase phases.StablePhase, manifestHash string) bool {
	for _, v := range phase.Cut {
		if v == manifestHash {
			return true
		}
	}
	return false
}

// blockUnpinned renders the refusal of a manifest the phase cut does not pin
// (the closed OUT_OF_SCOPE code — the DP03/DP04 refusal discipline — with the
// DP05 phase-authority explanation).
func blockUnpinned(phaseVersion, manifestHash string) blockreason.BlockReason {
	return blockreason.BlockReason{
		Code:     blockreason.CodeOutOfScope,
		Severity: blockreason.SeverityBlocking,
		Explanation: "Émission refusée : le StackManifest (" + manifestHash + ") n'est PAS épinglé dans la coupe " +
			"de la phase (" + phaseVersion + "). La phase est la source AUTORITATIVE de l'émission (DP05, KRD §43) : " +
			"EmitStack ré-émet exactement ce que la phase a sélectionné — jamais un manifest hors-coupe (honnêteté ; " +
			"S96 « ré-émet depuis la phase »).",
		HowToFix: []string{
			"pin_the_manifest_in_the_phase : faites entrer le manifest dans la coupe — gravez sa version dans le Kernel (idée → miroir → /goal → approbation) puis recalculez la phase stable (S23, phases.IsStable).",
			"emit_from_the_pinned_version : si la phase épingle une AUTRE version du manifest, émettez CELLE-LÀ — la coupe décide, pas l'appelant.",
			"re_emit : relancez EmitStack une fois le manifest épinglé ; même phase content-adressée → mêmes octets sur toute machine.",
		},
	}
}

// envVarName folds a service name to its env-var stem (the DP04 fold,
// duplicated as a 4-line pure function because envemit keeps it unexported —
// coherence is proven by the crossed property mirror, not by import).
func envVarName(serviceName string) string {
	up := strings.ToUpper(serviceName)
	return strings.Map(func(r rune) rune {
		if (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') {
			return r
		}
		return '_'
	}, up)
}

// renderTraefikDynamic renders the traefik DYNAMIC config (file provider) —
// the projection of the SAME declared topology the DP03 labels encode: one
// HTTPS router (websecure + certresolver reference) + one HTTP→HTTPS redirect
// router per role=server service, the shared redirect middleware, and the
// docker_internal service URLs (http://<container>:<internal_port> — the
// container name is the ${APP_NAME} reference, never a hardcoded host).
// Hosts, names and the cert resolver stay ${VAR} REFERENCES resolved at
// deploy time (SPEC-stack-2026: no hardcoded URL/secret). Servers render in
// stable (sorted) order; newlines are "\n".
func renderTraefikDynamic(m stackmanifest.StackManifest, sourceHash string) []byte {
	sorted := make([]stackmanifest.Service, len(m.Services))
	copy(sorted, m.Services)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Name < sorted[j].Name })
	primary := ""
	servers := []stackmanifest.Service{}
	for _, s := range sorted {
		if s.Role == stackmanifest.RoleServer {
			if primary == "" {
				primary = s.Name
			}
			servers = append(servers, s)
		}
	}

	routers := []string{}
	services := []string{}
	for _, svc := range servers {
		router := "${APP_NAME}"
		host := "${APP_SUBDOMAIN}.${DOMAIN}"
		container := "${APP_NAME}"
		if svc.Name != primary {
			router = "${APP_NAME}-" + svc.Name
			host = svc.Name + ".${APP_SUBDOMAIN}.${DOMAIN}"
			container = "${APP_NAME}-" + svc.Name
		}
		routers = append(routers,
			"    "+router+":",
			"      rule: Host(`"+host+"`)",
			"      entryPoints:",
			"        - websecure",
			"      service: "+router,
			"      tls:",
			"        certResolver: ${CERT_RESOLVER_NAME}",
			"    "+router+"-http:",
			"      rule: Host(`"+host+"`)",
			"      entryPoints:",
			"        - web",
			"      middlewares:",
			"        - ${APP_NAME}-https-redirect",
			"      service: "+router,
		)
		services = append(services,
			"    "+router+":",
			"      loadBalancer:",
			"        servers:",
			// docker_internal: the container name on the shared network.
			"          - url: http://"+container+":"+strconv.Itoa(svc.InternalPort),
		)
	}

	lines := []string{
		"# " + generators.ProtectedMarker + ". source: " + sourceHash,
		"# Config dynamique Traefik (file provider) — projection des labels DP03 du même manifest.",
		"# Toutes les valeurs sont des références d'environnement résolues au déploiement — jamais en dur.",
		"http:",
		"  routers:",
	}
	lines = append(lines, routers...)
	lines = append(lines,
		"  middlewares:",
		"    ${APP_NAME}-https-redirect:",
		"      redirectScheme:",
		"        scheme: https",
		"  services:",
	)
	lines = append(lines, services...)
	return []byte(strings.Join(lines, "\n") + "\n")
}

// firstHandEdit refuses on the FIRST ledger-tracked disk file whose bytes no
// longer hash to the recorded output_hash — sorted-path order (deterministic:
// the same drifted set always refuses on the same path), generators.Drifted
// (S34/S78 law) as the single drift definition. An untracked file is not a
// drift (the emitter never recorded it).
func firstHandEdit(disk []regen.DiskFile, ledger []regen.LedgerEntry) *blockreason.BlockReason {
	byPath := make(map[string]regen.LedgerEntry, len(ledger))
	for _, e := range ledger {
		byPath[e.Path] = e // duplicate paths: the LAST wins (append-only head)
	}
	sorted := make([]regen.DiskFile, len(disk))
	copy(sorted, disk)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Path < sorted[j].Path })
	for _, f := range sorted {
		entry, tracked := byPath[f.Path]
		if !tracked {
			continue
		}
		if generators.Drifted(entry.OutputHash, f.Bytes) {
			br := blockreason.For(blockreason.CodeEmittedFileHandEdited)
			return &br
		}
	}
	return nil
}

// bundleHash folds the bundle's identity into ONE output content address:
// records.Hash over the canonical "phase\npath:output_hash\n…" summary (S02
// reused — never a forked hash scheme). Same phase → same artifacts → same
// BundleHash: the «ré-émettre deux fois, hash égaux» proof is one comparison.
func bundleHash(phaseVersion string, arts []composeemit.Artifact) string {
	lines := []string{"phase:" + phaseVersion}
	for _, a := range arts {
		lines = append(lines, a.Path+":"+a.OutputHash)
	}
	return records.Hash([]byte(strings.Join(lines, "\n") + "\n"))
}

// EmitStack is the DP05 composition — the COMPLETE stack emission of a phase:
//
//  1. the phase must be STABLE (S23 computed verdict; PHASE_NOT_STABLE
//     otherwise — deploy/preview are keyed on stable phases only);
//  2. the manifest must be PINNED in the phase cut (the phase is the
//     authoritative source; OUT_OF_SCOPE otherwise);
//  3. NO tracked gen/ file may be hand-edited (the S78 drift law; the first
//     drift refuses the WHOLE bundle with EMITTED_FILE_HAND_EDITED,
//     fail-closed BEFORE any emission — §9 anti-overwrite);
//  4. then it COMPOSES composeemit.Emit (DP03) + envemit.Emit (DP04) + the
//     traefik dynamic projection, and triple-addresses the result
//     (PhaseVersion S23 · SourceHash S02 · BundleHash).
//
// PURE: no DB, no clock, no RNG, no I/O — ledger and disk are explicit inputs
// (the MCP/CLI seam reads the files, the S78 pattern). Same phase → same
// bytes on every machine: the cornerstone mirror of the provisioning epic.
func EmitStack(
	phase phases.StablePhase,
	m stackmanifest.StackManifest,
	ledger []regen.LedgerEntry,
	disk []regen.DiskFile,
) (Bundle, *blockreason.BlockReason) {
	// (1) stable phases only — the verdict is COMPUTED (S23), never declared.
	if !phase.Stable {
		br := blockreason.For(blockreason.CodePhaseNotStable)
		return Bundle{}, &br
	}

	// (2) the phase pins the manifest — an invalid manifest refuses with the
	// DP02 closed code via the DP03 emitter below; here we address it first.
	manifestHash, err := stackmanifest.HashManifest(m)
	if err != nil {
		// Reuse the DP03 refusal verbatim: emit and let it carry the DP02 code.
		_, br := composeemit.Emit(m)
		return Bundle{}, br
	}
	phaseVersion, verr := phase.Version()
	if verr != nil {
		// A phase that cannot canonicalize is not a source (defensive: phases
		// built by IsStable always canonicalize).
		br := blockreason.For(blockreason.CodePhaseNotStable)
		return Bundle{}, &br
	}
	if !manifestPinned(phase, manifestHash) {
		br := blockUnpinned(phaseVersion, manifestHash)
		return Bundle{}, &br
	}

	// (3) the hand-edit gate, BEFORE any emission (fail-closed, §9).
	if br := firstHandEdit(disk, ledger); br != nil {
		return Bundle{}, br
	}

	// (4) compose DP03 + DP04 + the traefik dynamic projection.
	compose, br := composeemit.Emit(m)
	if br != nil {
		return Bundle{}, br
	}
	env, br := envemit.Emit(m)
	if br != nil {
		return Bundle{}, br
	}
	traefikBytes := renderTraefikDynamic(m, manifestHash)
	traefik := composeemit.Artifact{
		Path:       "back/gen/" + m.AppName + "/traefik.dynamic.yml",
		Target:     TargetTraefikDynamic,
		Kind:       KindStackManifest,
		Bytes:      traefikBytes,
		SourceHash: manifestHash,
		OutputHash: records.Hash(traefikBytes),
		Protected:  true,
	}

	b := Bundle{
		PhaseVersion:     phaseVersion,
		SourceHash:       manifestHash,
		Compose:          compose,
		EnvExample:       env.EnvExample,
		StartSh:          env.StartSh,
		StartWithRebuild: env.StartWithRebuild,
		TraefikDynamic:   traefik,
	}
	b.BundleHash = bundleHash(phaseVersion, b.Artifacts())
	return b, nil
}

File: front/web/lib/stack-emit.ts
/**
 * stack-emit.ts — the PURE TS twin of the authoritative Go emitter
 * `back/runtime/composeemit` (DP03 — the additive Target docker-compose:
 * Emit(stack_manifest) → docker-compose.yml, /data/dockers conventions).
 *
 * DETERMINISM-FIRST (CLAUDE.md §6/§8, ADR 0036 EMITTED_FUNCTION_PURE): the
 * emission is a pure function of Canonicalize(manifest) — no clock, no RNG,
 * services and volumes in stable (sorted) order, "\n" newlines. THE GO CODE
 * IS AUTHORITATIVE; this twin is byte-parity-pinned to it by the vitest
 * mirror (stack-emit.test.ts pins GO_OUTPUT_HASH — the sha-256 of the full
 * Go-emitted bytes for the seeded Example). Any one-byte divergence reds.
 *
 * The conventions reproduced (boilerplates/{nodejs,payload}, alphashop):
 * container_name ${APP_NAME} (primary server) / ${APP_NAME}-<svc>;
 * env_file .env; restart: unless-stopped; NO published ports (Traefik labels
 * only — HTTPS router websecure/tls + HTTP→HTTPS redirect middleware); named
 * bind volumes driver_opts {type:none, device:${<DEVICE_VAR>}, o:bind};
 * traefik_default external:true except the deployment that OWNS it.
 *
 * THE WALL (CLAUDE.md §2): emitting is a below-the-line projection — this
 * module writes nothing; the /stack-emit route only MEASURES (validate +
 * emit + hash). The StackManifest source stays above the line (DP02).
 *
 * ADR 0043: the PRIMARY infra artifact is the Pulumi/TS program
 * (TargetPulumiProgram); this compose is the /data/dockers-convention
 * projection of the same StackManifest — deployable today, regenerable,
 * never the source of truth.
 */

import { validateDatastore } from "./environments";
import {
	hashManifest,
	PROFILES,
	type Profile,
	type Refusal,
	type Service,
	type StackManifest,
	serviceInProfile,
	validate,
} from "./stack-manifest";

/** The DP03 additive target of the closed kind × target matrix. */
export const TARGET_DOCKER_COMPOSE = "docker-compose";

/** One emitted projection — double content-addressed (source + output). */
export interface ComposeArtifact {
	path: string;
	target: typeof TARGET_DOCKER_COMPOSE;
	kind: "stack_manifest";
	yaml: string;
	sourceHash: string;
	outputHash: string;
	protected: true;
}

async function sha256hex(s: string): Promise<string> {
	const { createHash } = await import("node:crypto");
	return createHash("sha256").update(s, "utf8").digest("hex");
}

function sortedServices(m: StackManifest): Service[] {
	return [...m.services].sort((a, b) => (a.name < b.name ? -1 : 1));
}

function primaryServer(sorted: Service[]): string {
	for (const s of sorted) {
		if (s.role === "server") return s.name;
	}
	return "";
}

/** ${<SERVICE>_IMAGE} — the env-var image reference of an EMITTED service. */
function envVarImage(serviceName: string): string {
	const mapped = serviceName.toUpperCase().replace(/[^A-Z0-9]/g, "_");
	return `\${${mapped}_IMAGE}`;
}

// biome-ignore-start lint/suspicious/noTemplateCurlyInString: literal compose ${VAR} placeholders, interpolated by docker compose (the /data/dockers env-var convention — never a hardcoded value)
function renderTraefikLabels(svc: Service, primary: string): string[] {
	const router =
		svc.name === primary ? "${APP_NAME}" : `\${APP_NAME}-${svc.name}`;
	const host =
		svc.name === primary
			? "${APP_SUBDOMAIN}.${DOMAIN}"
			: `${svc.name}.\${APP_SUBDOMAIN}.\${DOMAIN}`;
	return [
		"    labels:",
		'      - "traefik.enable=true"',
		`      - "traefik.http.routers.${router}.rule=Host(\`${host}\`)"`,
		`      - "traefik.http.routers.${router}.entrypoints=websecure"`,
		`      - "traefik.http.routers.${router}.tls=true"`,
		`      - "traefik.http.routers.${router}.tls.certresolver=\${CERT_RESOLVER_NAME}"`,
		`      - "traefik.http.services.${router}.loadbalancer.server.port=${svc.internal_port}"`,
		`      - "traefik.http.routers.${router}-http.rule=Host(\`${host}\`)"`,
		`      - "traefik.http.routers.${router}-http.entrypoints=web"`,
		`      - "traefik.http.routers.${router}-http.middlewares=\${APP_NAME}-https-redirect"`,
		'      - "traefik.http.middlewares.${APP_NAME}-https-redirect.redirectscheme.scheme=https"',
	];
}

function renderHealthcheck(cmd: string): string[] {
	return [
		"    healthcheck:",
		`      test: ["CMD-SHELL", ${JSON.stringify(cmd)}]`,
		"      interval: 30s",
		"      timeout: 3s",
		"      retries: 3",
		"      start_period: 10s",
	];
}

function renderService(
	svc: Service,
	m: StackManifest,
	primary: string,
): string[] {
	const lines = [`  ${svc.name}:`];
	const container =
		svc.name === primary ? "${APP_NAME}" : `\${APP_NAME}-${svc.name}`;
	lines.push(`    container_name: ${container}`);
	const image =
		svc.image && svc.image !== "" ? svc.image : envVarImage(svc.name);
	lines.push(
		`    image: ${image}`,
		"    env_file:",
		"      - .env",
		"    restart: unless-stopped",
	);
	if (svc.profile !== "core") {
		lines.push("    profiles:", `      - ${svc.profile}`);
	}
	if (svc.depends_on && svc.depends_on.length > 0) {
		lines.push("    depends_on:");
		for (const d of svc.depends_on) {
			lines.push(`      - ${d}`);
		}
	}
	lines.push("    networks:", `      - ${m.network.name}`);
	if (svc.healthcheck && svc.healthcheck !== "") {
		lines.push(...renderHealthcheck(svc.healthcheck));
	}
	if (svc.role === "server") {
		lines.push(...renderTraefikLabels(svc, primary));
	}
	return lines;
}

function renderNetworks(m: StackManifest): string[] {
	const lines = ["networks:", `  ${m.network.name}:`];
	if (m.network.external) {
		lines.push("    external: true");
	}
	lines.push("    name: ${TRAEFIK_NETWORK_NAME}");
	return lines;
}

function renderVolumes(m: StackManifest): string[] | null {
	const vols = [...(m.volumes ?? [])].sort((a, b) =>
		a.name < b.name ? -1 : 1,
	);
	if (vols.length === 0) return null;
	const lines = ["volumes:"];
	for (const v of vols) {
		const name = vols.length > 1 ? `\${APP_NAME}-${v.name}` : "${APP_NAME}";
		lines.push(
			`  ${v.name}:`,
			`    name: ${name}`,
			"    driver: local",
			"    driver_opts:",
			"      type: none",
			`      device: \${${v.device_var}}`,
			"      o: bind",
		);
	}
	return lines;
}

// biome-ignore-end lint/suspicious/noTemplateCurlyInString: literal compose ${VAR} placeholders

/**
 * renderCompose — the pure renderer (the twin of Go composeemit.render):
 * protected header, services in stable order, networks, volumes.
 * Exported for the ∀ mirror; emitCompose is the validating door.
 */
export function renderCompose(m: StackManifest, sourceHash: string): string {
	const sorted = sortedServices(m);
	const primary = primaryServer(sorted);
	const sections = [
		`# CODE GENERATED BY AIDOS — DO NOT EDIT. source: ${sourceHash}`,
		"services:",
	];
	for (const svc of sorted) {
		sections.push(renderService(svc, m, primary).join("\n"));
	}
	sections.push("", renderNetworks(m).join("\n"));
	const vols = renderVolumes(m);
	if (vols) {
		sections.push("", vols.join("\n"));
	}
	return `${sections.join("\n")}\n`;
}

/**
 * emitCompose — the DP03 pure emitter (twin of Go composeemit.Emit):
 * validate (the DP02 closed codes — an invalid manifest is a Refusal, never
 * a guessed compose), content-address the source (hashManifest — S02
 * reused), render, content-address the output.
 */
export async function emitCompose(
	m: StackManifest,
): Promise<ComposeArtifact | { refusal: Refusal }> {
	const refusal = validate(m);
	if (refusal) {
		return { refusal };
	}
	const sourceHash = await hashManifest(m);
	const yaml = renderCompose(m, sourceHash);
	return {
		path: `back/gen/${m.app}/docker-compose.yml`,
		target: TARGET_DOCKER_COMPOSE,
		kind: "stack_manifest",
		yaml,
		sourceHash,
		outputHash: await sha256hex(yaml),
		protected: true,
	};
}

/**
 * The DP11 BlockReason codes carried by FilterByProfile — the exact twin of the
 * Go back/runtime/blockreason additions (UNKNOWN_PROFILE + the delegated DP06
 * DOLTGRES_NOT_ALLOWED_IN_PROD). A selection outside the closed set is never
 * coerced to the nearest known profile, never silently widened to `full`.
 */
export type ProfileBlockCode =
	| "UNKNOWN_PROFILE"
	| "DOLTGRES_NOT_ALLOWED_IN_PROD";

export interface ProfileBlock {
	code: ProfileBlockCode;
	message: string;
}

/**
 * filterByProfile — the DP11 PURE include/exclude over the CLOSED SPEC-stack-2026
 * profile set, the exact twin of Go composeemit.FilterByProfile. It adds no
 * business rule — it SELECTS which services of an already-declared manifest enter
 * the emission, deterministically:
 *
 *  1. the selection MUST be a member of the closed set (UNKNOWN_PROFILE else —
 *     never guessed, never coerced to `full`);
 *  2. the cross `non-prod` × `prod` is refused via the EXISTING DP06 gate
 *     (validateDatastore(env, "doltgres") reused verbatim, never forked —
 *     DOLTGRES_NOT_ALLOWED_IN_PROD): the `non-prod` profile carries Doltgres and
 *     prod imposes Postgres (ADR 0065);
 *  3. it keeps EXACTLY the services for which serviceInProfile(svc, selection)
 *     holds — `full` keeps every service (the deterministic UNION), any other
 *     profile keeps its declared members plus the always-running core services.
 *
 * The kept services preserve the manifest's declared order; everything else of
 * the manifest is carried through unchanged. PURE: same (manifest, profile, env)
 * → same filtered manifest. THE WALL (§2): the profiles are DECLARED above the
 * line in stack_manifest; the SELECTION is applied below the line at emission —
 * this reads the AST and returns a narrowed AST, it writes no truth.
 */
export function filterByProfile(
	m: StackManifest,
	profile: string,
	env: string,
): StackManifest | { block: ProfileBlock } {
	// (1) the selection is a DECLARED member of the closed set — never guessed.
	if (!(PROFILES as readonly string[]).includes(profile)) {
		return {
			block: {
				code: "UNKNOWN_PROFILE",
				message: `profile "${profile}" is outside the closed set [${PROFILES.join(" ")}] — a selection is declared, never inferred (DP11)`,
			},
		};
	}
	const selection = profile as Profile;

	// (2) the cross non-prod × prod ⇒ Doltgres refused — DELEGATE to the existing
	// DP06 gate (never a forked rule): the `non-prod` profile carries Doltgres,
	// and prod imposes Postgres. Only the non-prod selection bites; any other
	// profile against prod passes (the gate is keyed on the doltgres datastore).
	if (selection === "non-prod") {
		const refusal = validateDatastore(env, "doltgres");
		if (refusal && refusal.code === "DOLTGRES_NOT_ALLOWED_IN_PROD") {
			return {
				block: {
					code: "DOLTGRES_NOT_ALLOWED_IN_PROD",
					message: refusal.message,
				},
			};
		}
	}

	// (3) the pure include/exclude over the closed set — declared order preserved,
	// the caller's slice never mutated (a copy).
	const kept = m.services.filter((svc) => serviceInProfile(svc, selection));
	return { ...m, services: kept };
}

/** isProfileBlock — narrow the filterByProfile result to its DP11 refusal. */
export function isProfileBlock(
	v: StackManifest | { block: ProfileBlock },
): v is { block: ProfileBlock } {
	return "block" in v;
}

/**
 * driftDetected — a hand-edited emission is rejected: the bytes on disk no
 * longer match the recorded output_hash (the gen/ protection, CLAUDE.md §9).
 */
export async function driftDetected(
	ledgerOutputHash: string,
	onDisk: string,
): Promise<boolean> {
	return (await sha256hex(onDisk)) !== ledgerOutputHash;
}
